oeps/oep-0008.rst

```diff
@@ -0,0 +1,374 @@
+
+  * Configuration file: ``tox.ini``
+
+Code Coverage
+^^^^^^^^^^^^^
+
+* `Coverage.py`_ for calculating code coverage
+
+  * Configuration file: ``.coveragerc``
+
+* Report coverage stats to `codecov`_ for trending and analysis
+
+.. _pip-tools: https://pypi.python.org/pypi/pip-tools
+.. _Pylint: https://www.pylint.org/
+.. _edx-lint: https://github.com/edx/edx-lint/
+.. _pycodestyle: http://pycodestyle.pycqa.org/
+.. _pydocstyle: http://www.pydocstyle.org/
+.. _pytest: http://doc.pytest.org/
+.. _Tox: https://tox.readthedocs.io/
+.. _Coverage.py: https://coverage.readthedocs.io/
+.. _codecov: https://codecov.io/
+
+
+JavaScript Language Specific
+----------------------------
+
+Code Quality
+^^^^^^^^^^^^
+
+* `ESLint`_ for static analysis in order to find problematic patterns or code
+  that doesn’t adhere to style guidelines
+
+  * `edX ESLint config`_  for generating ESLint configuration files that verify
+    edX best practices
+
+  * Configure in either ``package.json`` or ``.eslintrc.json``
+
+  * Specify exclusions in ``.eslintignore``
+
+
+Test Framework
+^^^^^^^^^^^^^^
+
+* `Jasmine`_ as the framework for testing JavaScript code
+
+  * `jasmine-jquery`_ matchers and fixture loader for Jasmine
+
+  * Jasmine-based helper classes from the testing folder of the `edX UI
+    Toolkit`_
+
+* `Karma`_ as the test runner for the JS tests
+
+  * We also use plugins to extend Karma. As this list could change, it is not
+    codified here. Please look at the ``devDependencies`` section of the
+    ``package.json`` file in existing edX repos to survey those that we are
+    currently using.
+
+Code Coverage
+^^^^^^^^^^^^^
+
+* `JSCover`_ for calculating code coverage
+
+.. _ESLint: http://eslint.org/docs/about/
+.. _edX ESLint config: https://github.com/edx/eslint-config-edx
+.. _Jasmine: http://jasmine.github.io/
+.. _jasmine-jquery: https://github.com/velesin/jasmine-jquery
+.. _Karma: https://karma-runner.github.io/
+.. _edX UI Toolkit: http://ui-toolkit.edx.org/
+.. _JSCover: http://tntim96.github.io/JSCover/
+
+
+Rejected Alternatives:
+----------------------
+
+`nose`_ / `nose2`_
+  The original choice in the edx-platform was to use nose
+  to extend unittest to make testing nicer and easier to understand. However,
+  nose has been in maintenance mode for the past several years and will likely
+  cease without a new person/team to take over maintainership. The project
+  maintainers themselves suggest using nose2 or pytest (or plain
+  unittest/unittest2). As we still want to use a framework that extends
+  unittest, the question then becomes whether edX standardize on nose2 or
+  pytest.
+
+  Note that edX developers have already begun using pytest in new repos,
+  having chosen it over nose2 for some or all of the reasons listed below.
+  Given this and lacking other barriers that would make it a bad choice, we
+  have decided to standardize on pytest.
+
+  Here are some of the reasons that we have chosen pytest:
```

Comment:
I'm not sure I buy this, TBH. We never even evaluated nose2. I think in this case we are setting as a best practice what everyone is already doing. Each of the "reasons why pytest" is true for `nose2` as far as I understand it based on their docs. The only reason listed here which makes sense to me here is "it's what the industry is doing." That can be a valid reason because of reduced cognitive load and whatnot. But we should just redact the reasoning section, IMO, and simply point out that `pytest` is our tool of choice, even those many (really, almost all) of our repos are using nose.